operator/builtin/input/file/file_test.go

```diff
@@ -681,65 +686,182 @@ func TestMultiCopyTruncateSlow(t *testing.T) {
+			ephemeralLines:  true,
+		},
+		{
+			name:            "Fast/Deletion/ExceedFingerprint",
+			totalLines:      300,
+			maxLinesPerFile: 100,
+			maxBackupFiles:  1,
+			writeInterval:   time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+			ephemeralLines:  true,
+		},
+		{
+			name:            "Slow/NoRotation",
+			totalLines:      10,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/NoDeletion",
+			totalLines:      20,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion",
+			totalLines:      30,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion/ExceedFingerprint",
+			totalLines:      300,
+			maxLinesPerFile: 100,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
 	}
 
-	waitForMessages(t, logReceived, expected)
-}
-
-func TestRapidRotate_CopyTruncate(t *testing.T) {
-	t.Skip()
-	getMessage := func(m int) string { return fmt.Sprintf("message %d", m) }
-
-	numMessages := 1000
+	testRun := func(tc rotationTest, copyTruncate bool) func(t *testing.T) {
+		return func(t *testing.T) {
+			operator, logReceived, tempDir := newTestFileOperator(t,
+				func(cfg *InputConfig) {
+					cfg.PollInterval = helper.Duration{tc.pollInterval}
+				},
+				func(out *testutil.FakeOutput) {
+					out.Received = make(chan *entry.Entry, tc.totalLines)
+				},
+			)
+			logger := getRotatingLogger(t, tempDir, tc.maxLinesPerFile, tc.maxBackupFiles, copyTruncate)
+
+			expected := make([]string, 0, tc.totalLines)
+			for i := 0; i < tc.totalLines; i++ {
+				expected = append(expected, fmt.Sprintf("this is a fifty char log entry with the number %3d", i))
+			}
 
-	operator, logReceived, tempDir := newTestFileOperator(t, nil)
+			require.NoError(t, operator.Start())
+			defer operator.Stop()
 
-	logger := getRotatingLogger(t, tempDir, 10, true)
+			for _, message := range expected {
+				logger.Writer().Write([]byte(message + "\n"))
+				time.Sleep(tc.writeInterval)
+			}
 
-	expected := make([]string, 0, numMessages)
+			received := make([]string, 0, tc.totalLines)
+		LOOP:
+			for {
+				select {
+				case e := <-logReceived:
+					received = append(received, e.Record.(string))
+				case <-time.After(250 * time.Millisecond):
+					break LOOP
+				}
+			}
 
-	for i := 0; i < numMessages; i++ {
-		expected = append(expected, getMessage(i))
+			if tc.ephemeralLines {
+				if !tc.expectEphemeralLines() {
+					// This is helpful for test development, and ensures the sample computation is used
+					t.Logf("Potentially unstable ephemerality expectation for test: %s", tc.name)
```

Comment:
Am I understanding this correctly that we will still fail the test when there are ephemeral lines? I'm a little uncomfortable having tests that we expect to fail occasionally

Reply:
No, the test passes because we're only checking for a subset.

The purpose of this log line is twofold. 
1) To call attention to the developer that they may want to reconsider whether this test needs to be marked as potentially having ephemeral lines. In other words, the calculate is pretty conservative, and if it thinks the test can be held to a higher standard, then it probably should be.
2) To codify how ephemerality can be reasoned about in a clear way, and to use that code for something at least marginally useful, but not in a way where it is absolutely depended upon.

Reply:
Aah, I see now. Nice, that's awesome